src/sources/nats.rs

```diff
@@ -0,0 +1,182 @@
+use crate::{
+    config::{log_schema, DataType, SourceConfig, SourceContext, SourceDescription},
+    event::{Event, Value},
+    internal_events::NatsEventReceived,
+    shutdown::ShutdownSignal,
+    Pipeline,
+};
+use bytes::Bytes;
+use futures::{pin_mut, stream, SinkExt, Stream, StreamExt};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Could not create Nats subscriber: {}", source))]
+    NatsCreateError { source: std::io::Error },
+    #[snafu(display("Could not subscribe to Nats topics: {}", source))]
+    NatsSubscribeError { source: std::io::Error },
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct NatsSourceConfig {
+    url: String,
+    name: String,
+    subject: String,
+    queue: Option<String>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<NatsSourceConfig>("nats")
+}
+
+impl_generate_config_from_default!(NatsSourceConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "nats")]
+impl SourceConfig for NatsSourceConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        let (connection, subscription) = create_subscription(self).await?;
+
+        Ok(Box::pin(nats_source(
+            connection,
+            subscription,
+            cx.shutdown,
+            cx.out,
+        )))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "nats"
+    }
+}
+
+impl NatsSourceConfig {
+    fn to_nats_options(&self) -> async_nats::Options {
+        // Set reconnect_buffer_size on the nats client to 0 bytes so that the
+        // client doesn't buffer internally (to avoid message loss).
+        async_nats::Options::new()
+            .with_name(&self.name)
+            .reconnect_buffer_size(0)
+    }
+
+    async fn connect(&self) -> crate::Result<async_nats::Connection> {
+        self.to_nats_options()
+            .connect(&self.url)
+            .await
+            .map_err(|e| e.into())
+    }
+}
+
+impl From<NatsSourceConfig> for async_nats::Options {
+    fn from(config: NatsSourceConfig) -> Self {
+        async_nats::Options::new()
+            .with_name(&config.name)
+            .reconnect_buffer_size(0)
+    }
+}
+
+fn get_subscription_stream(
+    subscription: async_nats::Subscription,
+) -> impl Stream<Item = async_nats::Message> {
+    stream::unfold(subscription, |subscription| async move {
+        match subscription.next().await {
+            Some(msg) => Some((msg, subscription)),
+            None => None,
+        }
+    })
+}
+
+async fn nats_source(
+    _connection: async_nats::Connection,
```

Comment:
Is this parameter necessary?

Reply:
I had this function take ownership of both the `async_nats::Connection` and the `async_nats::Subscription` so that the two structs would be dropped at the same time, when this function returns. Previously, when I omitted this parameter, I had an issue where the Subscription wasn't receiving any messages because the Connection had already been dropped (I think). Is there a better way of doing this? Or does this just need a comment explaining what is going on?

Reply:
Hmm, that is odd. I would have thought that holding onto `Subscription` would hold onto `Connection` such that it wouldn't be dropped.

At the least, a comment would be good as it stands out as-is.